// detype/dist/cli.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/cli-lib.ts
var import_node_fs3 = __toESM(require("fs"));
var import_node_path2 = __toESM(require("path"));

// src/transformFile.ts
var import_node_fs2 = __toESM(require("fs"));

// src/transform.ts
var import_core = require("@babel/core");
var import_node_fs = require("fs");
var import_prettier = require("prettier");
var import_compiler_sfc = require("@vue/compiler-sfc");
var import_template_ast_types = require("@vuedx/template-ast-types");
var import_preset_typescript = __toESM(require("@babel/preset-typescript"));
var import_node_path = require("path");
function getDefinePropsObject(content) {
  const matched = /\sprops:\s*\{/m.exec(content);
  if (matched) {
    const startContentIndex = matched.index + matched[0].length - 1;
    let leftBracketCount = 1;
    let endContentIndex = startContentIndex + 1;
    while (leftBracketCount) {
      if (content.charAt(endContentIndex) === "{") {
        leftBracketCount++;
      } else if (content.charAt(endContentIndex) === "}") {
        leftBracketCount--;
      }
      endContentIndex++;
    }
    return content.substring(startContentIndex, endContentIndex);
  }
  return "";
}
async function transform(code, fileName, options = {}) {
  const { prettierOptions, ...removeTypeOptions } = options;
  const originalCode = code;
  const originalFileName = fileName;
  let propsContent = "";
  let emitsContent = "";
  code = code.replaceAll("\r\n", "\n");
  if (fileName.endsWith(".vue")) {
    const parsedVue = (0, import_compiler_sfc.parse)(code);
    if (parsedVue.descriptor.script?.lang !== "ts" && parsedVue.descriptor.scriptSetup?.lang !== "ts") {
      return originalCode;
    }
    let { script: script1, scriptSetup: script2 } = parsedVue.descriptor;
    const isContainsDefinePropsType = script2?.content.match(/defineProps\s*</m);
    const isContainsDefineEmitType = script2?.content.match(/defineEmits\s*</m);
    if (isContainsDefinePropsType || isContainsDefineEmitType) {
      const typescript = await import("typescript");
      (0, import_compiler_sfc.registerTS)(() => typescript.default);
      const resolveFile = (file) => {
        return (0, import_node_path.resolve)(
          (0, import_node_path.isAbsolute)(file) || file.startsWith("node_modules") ? "" : (0, import_node_path.dirname)(fileName),
          file
        );
      };
      const { content } = (0, import_compiler_sfc.compileScript)(parsedVue.descriptor, {
        id: fileName,
        fs: {
          fileExists(file) {
            return !!(0, import_node_fs.statSync)(resolveFile(file), {
              throwIfNoEntry: false
            })?.isFile();
          },
          readFile: (file) => {
            return (0, import_node_fs.readFileSync)(resolveFile(file)).toString();
          }
        }
      });
      if (isContainsDefinePropsType) {
        propsContent = getDefinePropsObject(content);
      }
      if (isContainsDefineEmitType) {
        emitsContent = content.match(/\semits:\s(\[.*\]?)/m)?.[1] || "";
      }
    }
    if (script1 && script2 && script1.loc.start.offset < script2.loc.start.offset) {
      [script2, script1] = [script1, script2];
    }
    const removeVueSfcScriptOptions = {
      ...removeTypeOptions,
      customizeBabelConfig(config) {
        config.plugins ||= [];
        config.plugins?.push({
          name: "detype-remove-with-defaults",
          visitor: {
            CallExpression(path2) {
              const callee = path2.get("callee");
              if (callee.isIdentifier()) {
                if (callee.node.name === "defineProps") {
                  const parentPath = path2.parentPath;
                  if (parentPath.isCallExpression()) {
                    const callee2 = parentPath.get("callee");
                    if (callee2.isIdentifier() && callee2.node.name === "withDefaults") {
                      parentPath.replaceWith(path2.node);
                      parentPath.stop();
                    }
                  }
                }
              }
            }
          }
        });
        removeTypeOptions.customizeBabelConfig?.(config);
      }
    };
    code = await removeTypesFromVueSfcScript(
      code,
      fileName,
      script1,
      parsedVue.descriptor.template?.ast,
      removeVueSfcScriptOptions
    );
    code = await removeTypesFromVueSfcScript(
      code,
      fileName,
      script2,
      parsedVue.descriptor.template?.ast,
      removeVueSfcScriptOptions
    );
  } else {
    code = await removeTypes(code, fileName, removeTypeOptions);
  }
  if (propsContent) {
    code = code.replace("defineProps(", (str) => `${str}${propsContent}`);
  }
  if (emitsContent) {
    code = code.replace("defineEmits(", (str) => `${str}${emitsContent}`);
  }
  code = (0, import_prettier.format)(code, {
    ...prettierOptions,
    filepath: originalFileName
  });
  return code;
}
async function removeTypes(code, fileName, options) {
  code = code.replace(
    /\n\n+/g,
    (match) => `
/* @detype: empty-line=${match.length} */
`
  );
  code = processMagicComments(code);
  const removeComments = {
    enter(p) {
      if (!p.node.leadingComments)
        return;
      for (let i = p.node.leadingComments.length - 1; i >= 0; i--) {
        const comment = p.node.leadingComments[i];
        if (code.slice(comment.end).match(/^\s*\n\s*\n/) || comment.value.includes("@detype: empty-line")) {
          break;
        }
        comment.value = "@detype: remove-me";
      }
    }
  };
  const babelConfig = {
    filename: fileName,
    retainLines: true,
    plugins: [
      // Plugin to remove leading comments attached to TypeScript-only constructs
      {
        name: "detype-comment-remover",
        visitor: {
          TSTypeAliasDeclaration: removeComments,
          TSInterfaceDeclaration: removeComments,
          TSDeclareFunction: removeComments,
          TSDeclareMethod: removeComments,
          TSImportType: removeComments
        }
      }
    ].filter(Boolean),
    presets: [import_preset_typescript.default],
    generatorOpts: {
      shouldPrintComment: (comment) => comment !== "@detype: remove-me" && (!options.removeTsComments || !comment.match(/^\s*(@ts-ignore|@ts-expect-error)/))
    }
  };
  if (options.customizeBabelConfig) {
    options.customizeBabelConfig(babelConfig);
  }
  const babelOutput = await (0, import_core.transformAsync)(code, babelConfig);
  if (!babelOutput || babelOutput.code === void 0 || babelOutput.code === null) {
    throw new Error("Babel error");
  }
  return babelOutput.code.replaceAll(/\n\n*/g, "\n").replace(
    /\/\* @detype: empty-line=([0-9]+) \*\//g,
    (_match, p1) => `
`.repeat(p1 - 2)
  );
}
async function removeTypesFromVueSfcScript(code, fileName, script, templateAst, options) {
  if (script === null || script.lang !== "ts")
    return code;
  if (script.setup && templateAst) {
    const expressions = /* @__PURE__ */ new Set();
    (0, import_template_ast_types.traverse)(templateAst, {
      enter(node) {
        let content = "";
        if ((0, import_template_ast_types.isSimpleExpressionNode)(node) && !node.isStatic) {
          const ForOfOrInRE = /\s+(of|in)\s+/;
          if (node.content.match(ForOfOrInRE)) {
            const parts = node.content.split(ForOfOrInRE);
            if (parts.length === 3) {
              content = parts[parts.length - 1];
            } else {
              content = node.content;
            }
          } else {
            content = node.content;
          }
        } else if ((0, import_template_ast_types.isComponentNode)(node)) {
          content = node.tag;
        }
        if (content) {
          expressions.add(`[${content}]`);
        }
      }
    });
    script.content += "/* @detype: remove-after-this */" + [...expressions].join(";");
  }
  let scriptCode = await removeTypes(script.content, fileName + ".ts", options);
  const removeAfterIndex = scriptCode.indexOf(
    "/* @detype: remove-after-this */"
  );
  if (removeAfterIndex >= 0) {
    scriptCode = scriptCode.slice(0, removeAfterIndex);
  }
  let before = code.slice(0, script.loc.start.offset);
  const after = code.slice(script.loc.end.offset);
  const matches = before.match(/\blang\s*=\s*["']ts["']/);
  if (matches) {
    const lastMatch = matches[matches.length - 1];
    const lastMatchIndex = before.lastIndexOf(lastMatch);
    before = before.slice(0, lastMatchIndex) + before.slice(lastMatchIndex + lastMatch.length);
  }
  return before + scriptCode + after;
}
function processMagicComments(input) {
  const REPLACE_COMMENT = "// @detype: replace\n";
  const WITH_COMMENT = "// @detype: with\n";
  const END_COMMENT = "// @detype: end\n";
  let start = input.indexOf(REPLACE_COMMENT);
  while (start >= 0) {
    const middle = input.indexOf(WITH_COMMENT, start);
    if (middle < 0)
      return input;
    const middleEnd = middle + WITH_COMMENT.length;
    const end = input.indexOf(END_COMMENT, middleEnd);
    if (end < 0)
      return input;
    const endEnd = end + END_COMMENT.length;
    const before = input.slice(0, start);
    const newText = input.slice(middleEnd, end).replaceAll(/^\s*\/\//gm, "");
    const after = input.slice(endEnd);
    input = before + newText + after;
    start = input.indexOf(REPLACE_COMMENT, before.length + newText.length);
  }
  return input;
}
function removeMagicComments(code, fileName, prettierOptions) {
  const REPLACE_COMMENT = "// @detype: replace\n";
  const WITH_COMMENT = "// @detype: with\n";
  const END_COMMENT = "// @detype: end\n";
  let start = code.indexOf(REPLACE_COMMENT);
  let startEnd = start + REPLACE_COMMENT.length;
  while (start >= 0) {
    const middle = code.indexOf(WITH_COMMENT, start);
    if (middle < 0)
      return code;
    const middleEnd = middle + WITH_COMMENT.length;
    const end = code.indexOf(END_COMMENT, middleEnd);
    if (end < 0)
      return code;
    const endEnd = end + END_COMMENT.length;
    const before = code.slice(0, start);
    const keptText = code.slice(startEnd, middle);
    const after = code.slice(endEnd);
    code = before + keptText + after;
    start = code.indexOf(REPLACE_COMMENT, before.length + keptText.length);
    startEnd = start + REPLACE_COMMENT.length;
  }
  code = (0, import_prettier.format)(code, {
    ...prettierOptions,
    filepath: fileName
  });
  return code;
}

// src/transformFile.ts
var import_prettier2 = require("prettier");
var { readFile, writeFile } = import_node_fs2.default.promises;
async function transformFile(inputFileName, outputFileName, options = {}) {
  const code = await readFile(inputFileName, "utf-8");
  const prettierOptions = await (0, import_prettier2.resolveConfig)(inputFileName);
  const output = await transform(code, inputFileName, {
    prettierOptions,
    ...options
  });
  await writeFile(outputFileName, output, "utf-8");
}
async function removeMagicCommentsFromFile(inputFileName, outputFileName) {
  const code = await readFile(inputFileName, "utf-8");
  const prettierConfig = await (0, import_prettier2.resolveConfig)(inputFileName);
  const output = await removeMagicComments(code, inputFileName, prettierConfig);
  await writeFile(outputFileName, output, "utf-8");
}

// src/cli-lib.ts
var import_fast_glob = __toESM(require("fast-glob"));

// package.json
var package_default = {
  name: "detypes",
  version: "0.7.9",
  description: "Removes TypeScript type annotations but keeps the formatting",
  main: "dist/index.js",
  bin: "detype.js",
  packageManager: "pnpm@8.8.0",
  engines: {
    node: ">=18"
  },
  scripts: {
    prepack: "rimraf dist && pnpm build",
    build: "tsup",
    dev: "tsup --watch",
    cq: "pnpm run test:typecheck && pnpm run test:lint && pnpm run test:prettier",
    test: "run-p 'test:*'",
    "test:unit": "vitest run",
    "test:typecheck": "tsc -p tsconfig.json --noEmit",
    "test:lint": "eslint src --max-warnings 0",
    "test:prettier": "prettier --check test-files --ignore-unknown . '!pnpm-lock.yaml'",
    format: "prettier . --write --ignore-path .gitignore --ignore-unknown . '!pnpm-lock.yaml'",
    release: "bumpp package.json"
  },
  files: [
    "dist/**/*",
    "index.d.ts"
  ],
  dependencies: {
    "@babel/core": "^7.21.4",
    "@babel/preset-typescript": "^7.21.4",
    "@vue/compiler-dom": "^3.4.0",
    "@vue/compiler-sfc": "^3.4.0",
    "@vuedx/template-ast-types": "0.7.1",
    "fast-glob": "^3.2.12",
    prettier: "^2.8.7",
    typescript: "^5.2.2"
  },
  devDependencies: {
    "@babel/traverse": "^7.21.4",
    "@cyco130/eslint-config": "^3.0.2",
    "@types/babel__core": "^7.20.0",
    "@types/babel__traverse": "^7.18.3",
    "@types/node": "18.15.11",
    "@types/prettier": "^2.7.2",
    bumpp: "^9.2.0",
    eslint: "^8.37.0",
    "npm-run-all": "^4.1.5",
    rimraf: "^4.4.1",
    tsup: "^6.7.0",
    vitest: "0.34.4"
  },
  repository: {
    type: "git",
    url: "git+https://github.com/Dunqing/detype.git"
  },
  keywords: [
    "typescript",
    "formatting",
    "vue",
    "sfc"
  ],
  author: "Dunqing <dengqing0821@gmail.com>",
  license: "MIT",
  bugs: {
    url: "https://github.com/Dunqing/detype/issues"
  },
  homepage: "https://github.com/Dunqing/detype#readme"
};

// src/cli-lib.ts
var { stat, mkdir } = import_node_fs3.default.promises;
async function cli(...args2) {
  let dashDash = false;
  const params = [];
  const flags = [];
  for (const arg of args2) {
    if (arg === "--") {
      dashDash = true;
    } else if (dashDash || !arg.startsWith("-")) {
      params.push(arg);
    } else {
      flags.push(arg);
    }
  }
  if (params.length > 2) {
    console.error("Too many arguments");
    return false;
  }
  let [input, output] = params;
  if (params.length === 0 || flags.some((flag) => flag === "-h" || flag === "--help")) {
    printUsage();
    return params.length > 0;
  }
  if (flags.some((flag) => flag === "-v" || flag === "--version")) {
    console.log(VERSION);
    return true;
  }
  const removeMagic = flags.some(
    (flag) => flag === "-m" || flag === "--remove-magic-comments"
  );
  const removeTsComments = flags.some(
    (flag) => flag === "-t" || flag === "--remove-ts-comments"
  );
  if (removeMagic && removeTsComments) {
    console.warn(
      "--remove-ts-comments has no effect when --remove-magic-comments is used"
    );
  }
  if (!removeMagic) {
    [input, output] = args2;
  }
  const inputStat = await stat(input);
  if (inputStat.isDirectory()) {
    if (!output) {
      console.error("No output directory given");
      printUsage();
      return false;
    }
    const files = (await (0, import_fast_glob.default)(unixify(input + "/**/*.{ts,tsx,vue}"))).filter(
      (file) => !file.endsWith(".d.ts")
    );
    const dirs = [...new Set(files.map((file) => import_node_path2.default.dirname(file)))].sort();
    await mkdir(import_node_path2.default.normalize(output), { recursive: true });
    for (const dir of dirs) {
      const outDir = import_node_path2.default.join(output, import_node_path2.default.relative(input, dir));
      if (outDir === output)
        continue;
      await mkdir(import_node_path2.default.normalize(outDir), { recursive: true });
    }
    for (const file of files) {
      const inputDir = import_node_path2.default.dirname(import_node_path2.default.relative(input, file));
      const outputName = inferName(file, import_node_path2.default.join(output, inputDir));
      removeMagic ? await removeMagicCommentsFromFile(
        import_node_path2.default.normalize(file),
        import_node_path2.default.normalize(outputName)
      ) : await transformFile(
        import_node_path2.default.normalize(file),
        import_node_path2.default.normalize(outputName),
        { removeTsComments }
      );
    }
    return true;
  }
  if (output) {
    const outputStat = await stat(output).catch((error) => {
      if (error && error.code === "ENOENT") {
        return null;
      }
      throw error;
    });
    if (outputStat && outputStat.isDirectory()) {
      output = inferName(input, output);
    }
  } else {
    if (removeMagic) {
      console.error(
        "Output file name is required when removing magic comments"
      );
      return false;
    }
    if (input.endsWith(".vue")) {
      console.error("Output file name is required for .vue files");
      return false;
    }
    output = inferName(input);
  }
  const outputDir = import_node_path2.default.dirname(output);
  if (outputDir) {
    await mkdir(import_node_path2.default.normalize(outputDir), { recursive: true });
  }
  removeMagic ? await removeMagicCommentsFromFile(
    import_node_path2.default.normalize(input),
    import_node_path2.default.normalize(output)
  ) : await transformFile(import_node_path2.default.normalize(input), import_node_path2.default.normalize(output), {
    removeTsComments
  });
  return true;
  function inferName(input2, outputDir2) {
    let output2;
    const { dir, name, ext } = import_node_path2.default.parse(input2);
    if (removeMagic) {
      output2 = import_node_path2.default.join(outputDir2 ?? dir, `${name}${ext}`);
    } else if (ext === ".ts") {
      output2 = import_node_path2.default.join(outputDir2 ?? dir, name + ".js");
    } else if (ext === ".tsx") {
      output2 = import_node_path2.default.join(outputDir2 ?? dir, name + ".jsx");
    } else if (ext === ".vue") {
      output2 = import_node_path2.default.join(outputDir2 ?? dir, name + ".vue");
    } else {
      throw new Error(`Unknwon file extension ${input2}`);
    }
    return output2;
  }
}
function printUsage() {
  console.error(USAGE);
}
var USAGE = `Usage:

  detypes [-m | --remove-magic-comments] <INPUT> [OUTPUT]

    INPUT   Input file or directory

    OUTPUT  Output file or directory
      (optional if it can be inferred and it won't overwrite the source file)

    -t, --remove-ts-comments
      Remove @ts-ignore and @ts-expect-error comments

    -m, --remove-magic-comments
      Remove magic comments only, don't perform ts > js transform

  detypes [-v | --version]

    Print version and exit

  detypes [-h | --help]

    Print this help and exit`;
var VERSION = package_default.version;
function unixify(name) {
  return name.replaceAll(import_node_path2.default.sep, "/");
}

// src/cli.ts
var args = process.argv.slice(2);
cli(...args).then((success) => process.exit(success ? 0 : 1));
